tasktiger/task.py

```diff
@@ -1,30 +1,289 @@
+        scripts = self.tiger.scripts
+        _key = self.tiger._key
+
+        from_state = from_state or self.state
+        queue = self.queue
+
+        assert from_state
+        assert queue
+
+        scripts.fail_if_not_in_zset(_key(from_state, queue), self.id,
+                                    client=pipeline)
+        if to_state:
+            if not when:
+                when = time.time()
+            if mode:
+                scripts.zadd(_key(to_state, queue), when, self.id,
+                                  mode, client=pipeline)
+            else:
+                pipeline.zadd(_key(to_state, queue), self.id, when)
+            pipeline.sadd(_key(to_state), queue)
+        pipeline.zrem(_key(from_state, queue), self.id)
+
+        if not to_state: # Remove the task if necessary
+            if self.unique:
+                # Only delete if it's not in any other queue
+                check_states = set([ACTIVE, QUEUED, ERROR, SCHEDULED])
+                check_states.remove(from_state)
+                # TODO: Do the following two in one call.
+                scripts.delete_if_not_in_zsets(_key('task', self.id, 'executions'),
+                                                    self.id, [
+                    _key(state, queue) for state in check_states
+                ], client=pipeline)
+                scripts.delete_if_not_in_zsets(_key('task', self.id),
+                                                    self.id, [
+                    _key(state, queue) for state in check_states
+                ], client=pipeline)
+            else:
+                # Safe to remove
+                pipeline.delete(_key('task', self.id, 'executions'))
+                pipeline.delete(_key('task', self.id))
+
+        scripts.srem_if_not_exists(_key(from_state), queue,
+                _key(from_state, queue), client=pipeline)
+
+        if to_state == QUEUED:
+            pipeline.publish(_key('activity'), queue)
+
+        try:
+            pipeline.execute()
+        except redis.ResponseError as e:
+            if '<FAIL_IF_NOT_IN_ZSET>' in unicode(e):
+                raise TaskNotFound('Task {} not found in queue "{}" in state "{}".'.format(
+                    self.id, queue, from_state
+                ))
+            raise
+        else:
+            self._state = to_state
+
+    def execute(self):
+        func = self.func
+        if getattr(func, '_task_batch', False):
+            return func([{'args': self.args, 'kwargs': self.kwargs}])
+        else:
+            return func(*self.args, **self.kwargs)
+
+    def delay(self, when=None):
+        tiger = self.tiger
+
+        now = time.time()
+        self._data['time_last_queued'] = now
+        serialized_task = json.dumps(self._data)
+
+        # convert timedelta to datetime
+        if isinstance(when, datetime.timedelta):
+            when = datetime.datetime.utcnow() + when
+
+        # convert to unixtime
+        if isinstance(when, datetime.datetime):
```

Comment:
@wojcikstefan see https://github.com/closeio/tasktiger/commit/a55e972d8550174d57d66704c65e8e23d2262941

I decided against having an explicit isinstance check since this is not very pythonic.